tiledb/sm/filesystem/mem_filesystem.cc

```diff
@@ -0,0 +1,423 @@
+/**
+ * @file   mem_filesystem.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2020 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements the in-memory filesystem class
+ */
+
+#include <mutex>
+#include <sstream>
+#include <unordered_set>
+
+#include "tiledb/common/logger.h"
+#include "tiledb/sm/filesystem/mem_filesystem.h"
+
+using namespace tiledb::common;
+using namespace std;
+
+namespace tiledb {
+namespace sm {
+
+class MemFilesystem::FSNode {
+ public:
+  FSNode(const string& name)
+      : name_(name){};
+
+  virtual ~FSNode() = default;
+
+  /* Indicates if it's a file or a directory */
+  virtual bool is_dir() const = 0;
+
+  /* Lists the contents of a node */
+  virtual vector<string> ls(const string& full_path) const = 0;
+
+  /* Indicates if a given node is a child of this node */
+  virtual bool has_child(const string& child) const = 0;
+
+  /* A hashtable of all the next-level subnodes of this node*/
+  unordered_map<string, unique_ptr<FSNode>> children_;
+
+  /* The name of this node */
+  string name_;
+
+ protected:
+  /** Protects FSNode for reads/writes */
+  mutable mutex node_mutex_;
```

Comment:
I thought about this a little more and I think we should remove this node-mutex and put a mutex in the MemFilesystem. I'm thinking of scenarios like `remove` where we need to lock both the parent and child. So for now, let's just use a single mutex in the filesystem and lock all of its routines.

Reply:
Makes sense, I changed it to your suggestions